pkg/informer/informer.go

```diff
@@ -58,14 +61,28 @@ const (
 	byGroupVersionResourceIndex = "byGroupVersionResource"
 )
 
-// DynamicDiscoverySharedInformerFactory is a SharedInformerFactory that
+// GVRSource is a source fo information about available GVRs,
+// and provides a way to register for changes in the available GVRs (addition of removal)
+// so that the list of informers can be updated.
+type GVRSource interface {
+	ListGVRs() []schema.GroupVersionResource
+	GetGVRPartialMetadata(gvr schema.GroupVersionResource) (*GVRPartialMetadata, error)
+	Synced() bool
+	Register(notifyUpdateNeeded func())
+}
+
+// genericDynamicDiscoverySharedInformerFactory is a SharedInformerFactory that
```

Comment:
while we're re-naming this, it's not discovery based anymore, and "dynamic" shared informer factories already have a meaning upstream, so we should find a good name to expose that this manages the set of types itself over time

Reply:
Maybe `SlefDiscoveringSharedInformerFactory` or something, idk

Reply:
In my understanding, the fact that it was called Dynamic**Discovery**SharedInformerFactory is more because it implements the `discovery.ServerResourcesInterface` interface, but in a dynamic way (you can subscribe), so that you don't need to do some regular polling of the remote discovery client.
In this sense it is still a `DynamicDiscoverySharedInformerFactory`, whatever be the source of the GVR list it is based on (CRD informer or anything else)

Reply:
The origin of "discovery" in the name is because it used to use discovery to determine the available APIs. That was inefficient (scaled linearly with the # of workspaces) and we replaced it with the combination of known built-in types plus all the CRDs in the system.

Reply:
So I renamed the types this way:
- the cluster-aware one: `DiscoveringDynamicSharedInformerFactory` (it is a dedicated type since it has the additional `Cluster` method) - same as the old one.
- the generic one `GenericDiscoveringDynamicSharedInformerFactory`, serving for the cluster-unaware case, and as a basis for the cluster-aware case (above).

And the `NewXXX` methods are named this way:
- `NewGenericDiscoveringDynamicSharedInformerFactory()` creates a **generic** ddsif from a `GVRSource` and a `newInformer()` function (`newInformer` will be different according to the case: cluster-aware / cluster-unaware)
- `NewScopedDiscoveringDynamicSharedInformerFactory()` creates a **cluster-unaware** ddsif from a `gvrSource` and a dynamic client (`dynamic.Interface`)
- `NewDiscoveringDynamicSharedInformerFactory()` creates a **cluster-aware** ddsif (`DiscoveringDynamicSharedInformerFactory`) from a `GVRSource` and a dynamic cluster client (`kcpdynamic.ClusterInterface`)
- `NewSelfDiscoveringDynamicSharedInformerFactory() ` creates a **cluster-aware** ddsif (`DiscoveringDynamicSharedInformerFactory`) from a CRD informer and a REST config. It is the same as the old one.